### PR TITLE
Disable rocBLAS atomics when asked

### PR DESCRIPTION
**Pull request by jglaser**
For some reason, rocBLAS matrix ops (on AMD GPUs) are non-deterministic by default (as opposed to cuBLAS ones). Among other things, this causes covariance matrices to lose the positive semidefinite property.

This PR makes the rocBLAS backend opt-in to Tensorflow's determinism flag, restoring deterministic behavior.

I used the following code to test on gfx90a (requires jax, transformers, torch):

```
import jax.numpy as np
import jax
#jax.config.update('jax_platform_name', 'cpu')

from transformers import FlaxAutoModel, BertConfig

model_name = 'mossaic-candle/regex-gb-2021'
config = BertConfig.from_pretrained(model_name)
config.attention_probs_dropout_prob = 0.0
config.hidden_dropout_prob = 0.0
m = FlaxAutoModel.from_pretrained(model_name, config=config, from_pt=True)
m.params = m.to_fp32(m.params)
a = m(input_ids=np.array([[1],[0]]))['last_hidden_state']
b = m(input_ids=np.array([[0],[1]]))['last_hidden_state']
print(a[0] == b[1])
```

It should print an array with all elements equal to `True`  if the execution is deterministic. Execute with `TF_DETERMINISTIC_OPS=1` environment variable set.

```
pip list | grep jax
jax                          0.3.23
jaxlib                       0.3.23
```

### CODE REVIEW

**Comment by jayfurmanek** [comment]
FYI - the AMD ROCm CI is broken from before this PR.
We have a PR up to fix that: https://github.com/tensorflow/tensorflow/pull/58137

**Comment by gbaned** [comment]
Hi @cheshire Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @cheshire Can you please review this PR ? Thank you!

**Review by cheshire — APPROVED** [review]

File: tensorflow/compiler/xla/stream_executor/rocm/rocm_blas.cc [hunk 1]
```diff
@@ -43,6 +43,9 @@ limitations under the License.
 #include "tensorflow/compiler/xla/stream_executor/scratch_allocator.h"
 #include "tensorflow/compiler/xla/stream_executor/stream_executor.h"
 
+#include "tensorflow/core/util/determinism.h"
```
Comment (akuegel) [inline tensorflow/compiler/xla/stream_executor/rocm/rocm_blas.cc:46]: Please use the include tensorflow/tsl/util/determinism.h
And change the BUILD dependency accordingly.

**Review by akuegel — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/xla/stream_executor/rocm/rocm_blas.cc (hunk 1 above)
Comment (jglaser) [reply tensorflow/compiler/xla/stream_executor/rocm/rocm_blas.cc]: I made those changes -- pending CI

**Comment by jayfurmanek** [comment]
@rahulbatra85  FYI

**Review by akuegel — APPROVED** [review]

File: tensorflow/compiler/xla/stream_executor/rocm/rocm_blas.cc [hunk 2]
```diff
@@ -193,7 +196,19 @@ bool ROCMBlas::DoBlasInternalImpl(FuncT rocblas_func, Stream *stream,
   }
 
   gpu::ScopedActivateExecutorContext sac{parent_};
-  rocblas_status ret = rocblas_func(blas_, args...);
+
+  // set the atomics mode, leaving default to library
+  bool allow_atomics = !OpDeterminismRequired();
+  rocblas_status ret;
+  if (!allow_atomics) {
+    ret = rocblas_set_atomics_mode(blas_, rocblas_atomics_not_allowed);
```
[the remark this answers is not in the document]
Comment (jglaser) [reply tensorflow/compiler/xla/stream_executor/rocm/rocm_blas.cc:204]: This was exposed in ROCM 3.8
https://github.com/ROCmSoftwarePlatform/rocBLAS/blob/21f6c60b188b1d4bbe4e4fa78b80fb2e83848ce4/CHANGELOG.md#rocblas-2280-for-rocm-380

The current version is 5.4.0. Perhaps update your test rig? We could, alternatively, #ifdef the check conditional on the toolkit version. Do you have a preference?

[the remark this answers is not in the document]
Comment (akuegel) [reply tensorflow/compiler/xla/stream_executor/rocm/rocm_blas.cc:204]: We are running ROCM 5.3, but maybe we are missing this library? Here is the Dockerfile:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/ci_build/Dockerfile.rbe.rocm-ubuntu20.04-manylinux2014-multipython

[the remark this answers is not in the document]
Comment (akuegel) [reply tensorflow/compiler/xla/stream_executor/rocm/rocm_blas.cc:204]: Actually, seems my roll forward to ROCM 5.3 didn't work out (not sure why), so we are still at ROCM 5.0.

[the remark this answers is not in the document]
Comment (jglaser) [reply tensorflow/compiler/xla/stream_executor/rocm/rocm_blas.cc]: ROCM 5.0 should be OK. However, I didn't pay attention to the fact that the error message looked more like a linker error, not a compiler error. I also didn't notice that the other API calls for rocblas go through a wrapper, most likely with the intent to support loading the rocblas library dynamically at runtime. Therefore, I have updated the call to `rocblas_set_atomics_mode` to also use this wrapper, hopefully that will resolve the problem. Could you please rerun the CI?

**Review by akuegel — APPROVED** [review]

**Pull request merged**